iot_config/src/helium_netids.rs

```diff
@@ -0,0 +1,291 @@
+            HeliumNetId::Type6 => TYPE_6_ID,
+        }
+    }
+
+    pub fn addr_range(&self) -> RangeInclusive<u32> {
+        match *self {
+            HeliumNetId::Type0 => TYPE_0_RANGE,
+            HeliumNetId::Type3 => TYPE_3_RANGE,
+            HeliumNetId::Type6 => TYPE_6_RANGE,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+pub trait AddressStore {
+    type Error;
+
+    async fn get_used_addrs(&self) -> Result<Vec<u32>, Self::Error>;
+    async fn claim_addrs(&self, new_addrs: &[u32]) -> Result<(), Self::Error>;
+    async fn release_addrs(&self, released_addrs: &[u32]) -> Result<(), Self::Error>;
+}
+
+#[async_trait::async_trait]
+impl AddressStore for sqlx::Pool<sqlx::Postgres> {
+    type Error = sqlx::Error;
+
+    async fn get_used_addrs<'a>(&'a self) -> Result<Vec<u32>, Self::Error> {
+        Ok(
+            sqlx::query_scalar::<_, i32>(" select devaddr from helium_used_devaddrs ")
+                .fetch_all(self)
+                .await?
+                .into_iter()
+                .map(|addr| addr as u32)
+                .collect::<Vec<u32>>(),
+        )
+    }
+
+    async fn claim_addrs<'a>(&'a self, new_addrs: &[u32]) -> Result<(), Self::Error> {
+        let mut query_builder: sqlx::QueryBuilder<sqlx::Postgres> =
+            sqlx::QueryBuilder::new(" insert into helium_used_devaddrs (devaddr) ");
+        query_builder.push_values(new_addrs, |mut builder, addr| {
+            builder.push_bind(*addr as i32);
+        });
+        Ok(query_builder.build().execute(self).await.map(|_| ())?)
+    }
+
+    async fn release_addrs<'a>(&'a self, released_addrs: &[u32]) -> Result<(), Self::Error> {
+        let mut query_builder: sqlx::QueryBuilder<sqlx::Postgres> =
+            sqlx::QueryBuilder::new(" delete from helium_used_devaddrs where devaddr in ");
+        query_builder.push_values(released_addrs, |mut builder, addr| {
+            builder.push_bind(*addr as i32);
+        });
+        Ok(query_builder.build().execute(self).await.map(|_| ())?)
+    }
+}
+
+pub async fn checkout_devaddr_constraints<S>(
+    addr_store: &S,
+    count: u64,
+    net_id: HeliumNetId,
+) -> Result<Vec<DevAddrConstraint>, DevAddrConstraintsError<S::Error>>
+where
+    S: AddressStore,
+{
+    let addr_range = net_id.addr_range();
+    let mut used_addrs = addr_store
+        .get_used_addrs()
+        .await
+        .map_err(DevAddrConstraintsError::AddressStore)?;
+    used_addrs.sort();
+    let range_start = *addr_range.start();
+    let range_end = *addr_range.end();
+    let last_used = if let Some(last) = used_addrs.last() {
+        *last
+    } else {
+        range_start
+    };
+    let used_range = (range_start..=last_used).collect::<HashSet<u32>>();
+    let used_addrs = used_addrs.into_iter().collect::<HashSet<u32>>();
+    let mut diff = used_range
+        .difference(&used_addrs)
+        .copied()
+        .collect::<Vec<_>>();
+    let mut next_addr = last_used + 1;
+    while diff.len() < count as usize {
```

Comment:
If there is an unallocated gap larger than the requested amount of devaddrs, the whole gap will be allocated.

```rust
    #[tokio::test]
    async fn allocate_fewer_than_existing_gap() {
        let used_addrs = Mutex::new(vec![]);
        checkout_devaddr_constraints(&used_addrs, 8, HeliumNetId::Type0)
            .await
            .expect("allocate first round");
        checkout_devaddr_constraints(&used_addrs, 32, HeliumNetId::Type0)
            .await
            .expect("allocate second round");
        checkout_devaddr_constraints(&used_addrs, 8, HeliumNetId::Type0)
            .await
            .expect("allocate third round");

        // round 2 goes out of business, and their devaddrs are released back to the wild
        let remove: Vec<u32> = used_addrs
            .lock()
            .await
            .clone()
            .into_iter()
            .skip(8)
            .take(32)
            .collect();
        assert_eq!(Ok(()), used_addrs.release_addrs(&remove).await);
        assert_eq!(8 + 8, used_addrs.lock().await.len());
        checkout_devaddr_constraints(&used_addrs, 8, HeliumNetId::Type0)
            .await
            .expect("allocate fourth round");
        assert_eq!(8 + 8 + 8, used_addrs.lock().await.len());
    }
```

Reply:
good catch! fixed [c6cca67](https://github.com/helium/oracles/pull/514/commits/c6cca67e1acb8c529c73920ea6e11d9ba437c6ee)